### ISSUE TITLE
feature request: test_case generators

### ISSUE THREAD

**Issue by jtmoon79** [frondeus/test-case#111]
tl;dr would be great if some primitive test_case generators were possible, e.g. `#[test_case(1; 3; "test cases 1 2 and 3")]` (made-up syntax).

### Problem

For my project, I have a large static vector [`DATA`](https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/0.1.44/src/data/datetime.rs#L2003-L2073) with complex entries. `DATA` length corresponds to a global constant `DATA_LEN`. Each `DATA` entry carries it's own built-in test case (only in test builds). Previously, I had created one test case and looped on all entries.

```rust
#[test]
fn test_DATA_entries() {
    for entry in DATA.iter() {
         assert(...)
    }
}
```

But I wanted more granular test cases, one test case for each entry in the `DATA`. So I added [`#[test_case]` declaration for each](https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/67f2579cd9c80ed26b97db0da50c2babcaba11f3/src/tests/datetime_tests.rs#L493-L570)
```rust
#[test_case(0)]
#[test_case(1)]
// ...
#[test_case(70)]
fn test_DATA_entry(index: usize) {
    let entry = &DATA[&index];
    assert(...)
}
```

However, _I have to manually update the listing of `#[test_case]` when `DATA` changes_.

### Solution

**It would be great if I could have `test_case` help me generate the test cases, `0` up to `DATA_LEN`.**

```rust
const DATA_LEN: usize = 70;

#[test_case(0; DATA_LEN; "a test case for each entry in DATA")]
fn test_DATA_entry(index: usize) {
    let entry = &DATA[&index];
    assert(...)
}
```
the code snippet would run 70 unique test cases.

#### Suggestions

I'm imagining some new formatting like
```rust
#[test_case(1; 3; "test cases 1 2 and 3")]
```

Or perhaps a new `test_case` library export like
```rust
#[test_case_generator(1; 3; "test cases 1 2 and 3")]
```

Or perhaps it has a start, stop, step that takes closures. The stop closure returns `Option<T>`. When `None` the test cases stop else pass `T` to the test case
```rust
#[test_case_generator(1, |x| if x > 3 { None } else { Some(x) }, |x| x + 1; "test cases 1 2 and 3")]
```

**Comment by jtmoon79** [comment]
Similar to #94
